### ISSUE TITLE
Panic in `clone_map.rs`

### ISSUE THREAD

**Issue by Lysxia** [creusot-rs/creusot#1299]
```rs
extern crate creusot_contracts;
use creusot_contracts::*;

#[variant(i)]
pub fn count_paths(i: usize) {
    (0..i).map(|j| count_paths(j));
}
```

```
thread 'rustc' panicked at creusot/src/backend/clone_map.rs:383:17:
assertion `left == right` failed
  left: 2
 right: 1
```

The mentioned line:

https://github.com/creusot-rs/creusot/blob/739b2d0ecb8103f7fba017c0b09346954951bcb4/creusot/src/backend/clone_map.rs#L381-L386

**Comment by jhjourdan** [comment]
Ok, the error message is not nice. But the problem here is that this is essentially mutually recursive functions (closure / function), which we do not support.

**Comment by arnaudgolfouse** [comment]
> Ok, the error message is not nice. But the problem here is that this is essentially mutually recursive functions (closure / function), which we do not support.

But we do (in programs)? E.g.
```
pub fn count_paths(i: usize) {
    (0..i).map(|j| closure(j));
}

fn closure(j: usize) { count_paths(j) }
```
Works. The error is caused by the presence of a closure.

**Comment by jhjourdan** [comment]
Right. We do not support it in the case the mutual recursion is between a closure and the function which contains it. That's clearly a limitation we need to work on.

One issue is that Coma would not support (yet?) spec inference in this case.
